include/sparrow/mp_utils.hpp

```diff
@@ -14,19 +14,130 @@
 
 #pragma once
 
+#include <type_traits>
 #include <iterator>
 
-namespace sparrow
+namespace sparrow::mpl
 {
-    namespace mpl
+    /// A sequence of types, used for meta-programming operations.
+    template< class... T >
+    struct typelist {};
+
+    /// @returns The count of types contained in a given `typelist`.
+    template< class... T >
+    constexpr std::size_t size(typelist<T...>) { return sizeof...(T); }
+
+    template< class L, template<class...> class U >
+    struct is_type_instance_of : std::false_type {};
+
+    template< template<class...> class L, template<class...> class U, class... T>
+        requires std::is_same_v< L<T...>, U<T...> >
+    struct is_type_instance_of< L<T...>, U > : std::true_type {};
+
+    /// `true` if `T` is a concrete type template instanciation of `U` which is a type template.
+    /// Example: is_type_instant_of_v< std::vector<int>, std::vector > == true
+    template< class T, template< class... > class U >
+    constexpr bool is_type_instance_of_v = is_type_instance_of<T, U>::value;
+
+    /// Matches any type which is an instance of `typelist`.
+    /// Examples:
+    ///     stati_assert(any_typelist< typelist<int, float, std::vector<int> > > == true);
+    ///     stati_assert(any_typelist< std::vector<int> > == false);
+    template< typename TList >
+    concept any_typelist = is_type_instance_of_v<TList, typelist>;
+
+    // TODO: document this!
+    template <class T, bool is_const>
+    struct constify
+        : std::conditional<is_const, const T, T>
+    {
+    };
+
+    template <class T, bool is_const>
+    using constify_t = typename constify<T, is_const>::type;
+
+    /// Checks that at least one type in the provided list of is making the provide predicate return `true`.
+    /// @returns `true` if for any type T in the type list L, `Predicate{}(typelist<T>) == true`.
+    ///          `false` otherwise or if the list is empty.
+    template< template<class...> class L, class Predicate, class... T>
+        requires any_typelist<L<T...>>
+    consteval
+    bool any_of(L<T...> list, Predicate predicate)
     {
-        template <class T, bool is_const>
-        struct constify
-            : std::conditional<is_const, const T, T>
+        return (predicate(typelist<T>{}) || ... || false);
+    }
+
+    /// Checks that every type in the provided list of is making the provide predicate return `true`.
+    /// @returns `true` if for every type T in the type list L, `Predicate{}(typelist<T>) == true`
+    ///          or if the list is empty; `false` otherwise.
+    template< template<class...> class L, class Predicate, class... T>
+        requires any_typelist<L<T...>>
+    consteval
+        bool all_of(L<T...> list, Predicate predicate)
+    {
+        return (predicate(typelist<T>{}) && ... && true);
+    }
+
+    /// Compile-time type predicate: `true` if the evaluated type is the same as `T`.
```

Comment:
From my understanding, this checks that the evaluated type (i.e. the type of `list` IIUC) is the same as `typelist<T>`.
Why do we need to compare typelists of types instead of types directly?

Reply:
It's a general "what is a type predicate" issue:
I tried to use values+functions instead of template types to do the meta functions, which makes difficult to decide how to pass a predicate that acts on a type. My current soltuion (which is undocumented but used everywhere here for now, and in tests) is that a type predicate is instanciated and the object is called with as argument a typelist with 1 type element. That way we can easilly deduce what is the type.

However, this is not conventional.

Ignoring the typedef-way of doing it, one could imagine something like:
```
any_of(typelist<X,Y>{}, std::same_as<T>); // no, because we can't pass concepts
any_of(typelist<X,Y>{}, std::is_same_as<???>); // can't be used as a predicate this way
any_of(typelist<X,Y>{}, same_as<T>); // works only if `same_as` is an object, which can work (like I did in the tests)
``` 
that last line assumes then that the algorithm is written in a way to pass each type, so how to do this?

```c++
template< template<class...> class L, class Predicate, class... T>
        requires any_typelist<L<T...>>
    consteval
    bool any_of(L<T...> list, Predicate predicate)
    {
        // series of attempts:
        return (predicate(T{}) || ... || false);  // not good because we dont know the consequences of creating+destryoing T
        return (predicate<T>() || ... || false);   // does not work syntactically because predicate is not a template        
        return (predicate.operator()<T>() || ... || false);   // might work, not sure, but inelegant (maybe possible to wrap in a special invokation function?

        return (predicate(typelist<T>{}) || ... || false);   // we "wrap" T in whatever allows to retrive it on call, that works but is unconventional
    }

```


I'm open to suggestions for alternatives. I would just like to avoid predicates built around recursive typedefs as we have the opportunity to do better with C++20. 
I'll take a look at `boost_ext/mp` recent C++20  metaprog library which is similar to mp11 but with C++20 in mind, maybe they use a better technique. Ideally if ther was a simple way to make predicates using a lambda, that would be perfect because they automatically are constexpr if the definition allows it.

Reply:
Thanks for the explanation. My initial assumption was to do something like `(predicate(T{}) || ... || false)`, but indeed that would call the constructor and destructor of `T`.

Reply:
Thinking more about this, a solution would be a special compile-time function that provides various ways to invoke a type-predicate depending on how it can be used. I'll experiment.

Reply:
@JohanMabille for some reason github didnt show me your answers immediately, my answere was not meant for yours.

Yes `identity` is nice for that, I'll consider maybe making that predicate invocation function handle this, `typelist<T>{}` and  some otehr possibilities like `predicate<T>()` and `predicate.operator()<T>()` when available.

Reply:
I think I setup a nice solution for type predicates, supporting also type-traits from the standard automatically for algorithms that add the overload that handle that (see `mpl::any_of` or `mpl::all_of` overloads as simple examples).